aggregator/handler/common/traffic_controller.go

```diff
@@ -0,0 +1,205 @@
+func NewTrafficControlOptions() TrafficControlOptions {
+	return &trafficControlOptions{
+		defaultEnabled: defaultDefaultDisabled,
+		initTimeout:    defaultInitTimeout,
+		instrumentOpts: instrument.NewOptions(),
+	}
+}
+
+func (opts *trafficControlOptions) SetStore(store kv.Store) TrafficControlOptions {
+	o := *opts
+	o.store = store
+	return &o
+}
+
+func (opts *trafficControlOptions) Store() kv.Store {
+	return opts.store
+}
+
+func (opts *trafficControlOptions) SetDefaultEnabled(value bool) TrafficControlOptions {
+	o := *opts
+	o.defaultEnabled = value
+	return &o
+}
+
+func (opts *trafficControlOptions) DefaultEnabled() bool {
+	return opts.defaultEnabled
+}
+
+func (opts *trafficControlOptions) SetRuntimeEnableKey(value string) TrafficControlOptions {
+	o := *opts
+	o.runtimeEnableKey = value
+	return &o
+}
+
+func (opts *trafficControlOptions) RuntimeEnableKey() string {
+	return opts.runtimeEnableKey
+}
+
+func (opts *trafficControlOptions) SetInitTimeout(value time.Duration) TrafficControlOptions {
+	o := *opts
+	o.initTimeout = value
+	return &o
+}
+
+func (opts *trafficControlOptions) InitTimeout() time.Duration {
+	return opts.initTimeout
+}
+
+func (opts *trafficControlOptions) SetInstrumentOptions(value instrument.Options) TrafficControlOptions {
+	o := *opts
+	o.instrumentOpts = value
+	return &o
+}
+
+func (opts *trafficControlOptions) InstrumentOptions() instrument.Options {
+	return opts.instrumentOpts
+}
+
+// TrafficController controls if traffic is enabled.
+type TrafficController struct {
+	enabled *atomic.Bool
+	value   watch.Value
+	opts    TrafficControlOptions
+}
+
+// NewTrafficController creates a new traffic controller.
+func NewTrafficController(opts TrafficControlOptions) *TrafficController {
+	enabled := atomic.NewBool(opts.DefaultEnabled())
+	iOpts := opts.InstrumentOptions()
+	newUpdatableFn := func() (watch.Updatable, error) {
+		w, err := opts.Store().Watch(opts.RuntimeEnableKey())
+		return w, err
+	}
+	getFn := func(updatable watch.Updatable) (interface{}, error) {
+		return updatable.(kv.ValueWatch).Get(), nil
+	}
+	processFn := func(update interface{}) error {
+		b, err := util.BoolFromValue(
+			update.(kv.Value),
+			opts.RuntimeEnableKey(),
+			opts.DefaultEnabled(),
+			util.NewOptions().SetLogger(iOpts.Logger()),
+		)
+		if err != nil {
+			return err
+		}
+		enabled.Store(b)
+		return nil
+	}
+	vOptions := watch.NewOptions().
+		SetInitWatchTimeout(opts.InitTimeout()).
+		SetInstrumentOptions(iOpts).
+		SetNewUpdatableFn(newUpdatableFn).
+		SetGetUpdateFn(getFn).
+		SetProcessFn(processFn)
+	return &TrafficController{
+		enabled: enabled,
+		value:   watch.NewValue(vOptions),
+		opts:    opts,
+	}
+}
+
+// Init initializes the traffic controller to watch the runtime updates.
+func (c *TrafficController) Init() error {
+	err := c.value.Watch()
+	if err == nil {
+		return nil
+	}
+	if _, ok := err.(watch.CreateWatchError); ok {
+		return err
+	}
+	return nil
```

Comment:
What other errors can we return here?

Reply:
You mean from the value.Watch()? It can also be `InitValueError`.

Reply:
I see, do you want to be safe here and just return the error? This is usually because something is wrong with KV, in which case relying on the default value may not be safe, so I'm not sure if you want to proceed here when the error occurs.